skills/dff_reasoning_skill/scenario/response.py

```diff
@@ -217,79 +256,158 @@ def response_with_chosen_api(ctx: Context, actor: Actor, *args, **kwargs) -> str
+                GENERATIVE_SERVICE_URL,
+                GENERATIVE_SERVICE_CONFIG,
+                GENERATIVE_TIMEOUT,
+                sending_variables,
+            )
+            response = hypotheses[0]
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            logger.exception(e)
+            response = None
+        logger.info(f"self reflexion: {response}")
+        step += 1
+        int_ctx.save_to_shared_memory(ctx, actor, step=step)
+        int_ctx.save_to_shared_memory(ctx, actor, self_reflexion=response)
+        return response
 
 
-def response_with_approved_api(ctx: Context, actor: Actor, *args, **kwargs) -> str:
+def final_answer(ctx: Context, actor: Actor, *args, **kwargs) -> str:
     if not ctx.validation:
         shared_memory = int_ctx.get_shared_memory(ctx, actor)
-        api2use = shared_memory.get("api2use", None)
-        timeout = api_conf[api2use]["timeout"]
-        signal.signal(signal.SIGALRM, timeout_handler)
-        signal.alarm(timeout)
+        subtask_results = shared_memory.get("subtask_results", {})
+        user_request = shared_memory.get("user_request", "")
+        prompt = f"""Today date is: {today}. USER REQUEST: {user_request}
+CONTEXT:
+{"---".join(list(subtask_results.values()))}
+YOUR TASK: given the information in the context, form a final answer to the user request"""
+        dialog_context = []
+        human_uttr_attributes = int_ctx.get_last_human_utterance(ctx, actor).get("attributes", {})
+        lm_service_kwargs = human_uttr_attributes.pop("lm_service_kwargs", None)
+        lm_service_kwargs = {} if lm_service_kwargs is None else lm_service_kwargs
+        envvars_to_send = ENVVARS_TO_SEND if len(ENVVARS_TO_SEND) else human_uttr_attributes.get("envvars_to_send", [])
+        sending_variables = compose_sending_variables(
+            lm_service_kwargs,
+            envvars_to_send,
+            **human_uttr_attributes,
+        )
         try:
-            response = globals()[f"{api2use}_response"](ctx, actor)
-        except Exception:
-            response = "Unfortunately, somthing went wrong with API"
-        signal.alarm(0)
+            hypotheses = send_request_to_prompted_generative_service(
+                dialog_context,
+                prompt,
+                GENERATIVE_SERVICE_URL,
+                GENERATIVE_SERVICE_CONFIG,
+                GENERATIVE_TIMEOUT,
+                sending_variables,
+            )
+            response = hypotheses[0]
+        except Exception as e:
+            sentry_sdk.capture_exception(e)
+            logger.exception(e)
+            response = None
+
+        subtask_results = {}
+        plan = []
+        step = 0
+        int_ctx.save_to_shared_memory(ctx, actor, subtask_results=subtask_results)
+        int_ctx.save_to_shared_memory(ctx, actor, plan=plan)
+        int_ctx.save_to_shared_memory(ctx, actor, step=step)
+        logger.info(f"final answer: {response}")
+        return response
```

Comment:
response which is in some cases a string with answer must NOT be a None because there will be problems in agent and will cause crash and restart of agent container